### ISSUE TITLE
Gh-ost breaks when using unsigned MEDIUMINT

### ISSUE THREAD

**Issue by serdarsahin** [github/gh-ost#353]
When using unsigned integer data types, binary log behavior changes if the inserted number is higher than the max signed integer value.

The original table
```
CREATE TABLE `test_table` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT,
  `column1` int(11) NOT NULL,
  `column2` int(11) NOT NULL,
  `column3` mediumint(8) unsigned NOT NULL,
  `column4` tinyint(3) unsigned NOT NULL,
  `column5` int(11) NOT NULL,
  `column6` int(11) NOT NULL,
  PRIMARY KEY (`id`),
  KEY `user_id_level_id` (`column1`,`column2`)
) ENGINE=InnoDB AUTO_INCREMENT=4647930180 DEFAULT CHARSET=utf8
```

The new schema;

```
CREATE TABLE `test_table` (
  `id` bigint(20) NOT NULL AUTO_INCREMENT,
  `column1` int(11) NOT NULL,
  `column2` smallint(5) unsigned NOT NULL,
  `column3` mediumint(8) unsigned NOT NULL,
  `column4` tinyint(3) unsigned NOT NULL,
  `column6` int(11) NOT NULL,
  PRIMARY KEY (`column1`,`column2`),
  UNIQUE KEY `id` (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=4632662231 DEFAULT CHARSET=utf8
```

We were experiencing the same issue explained on the issue #348 . Alter was never starting, even for hours. The table has more than 4 billion records and receiving 200 million records per week. So we tested the version that has included in this comment: https://github.com/github/gh-ost/issues/348#issuecomment-270592112 and it was quite fast. It was estimating 35 hours to complete which is perfectly acceptable for us.

We then executed the following statement

```
./gh-ost --max-load=Threads_running=800 --critical-load=Threads_running=1200 --chunk-size=50000 --dml-batch-size=100 --user="root" --password="" --host=localhost --database="test" --table="test_table" --verbose --alter="ADD UNIQUE INDEX (id), DROP COLUMN column5, DROP INDEX column1_column2, DROP PRIMARY KEY,ADD PRIMARY KEY(column1,column2),MODIFY column2 SMALLINT UNSIGNED NOT NULL, MODIFY column3 INT NOT NULL" --cut-over=default --exact-rowcount --concurrent-rowcount --default-retries=120 --panic-flag-file=ghost.panic.flag --postpone-cut-over-flag-file=ghost.postpone.flag
```

These are from binary logs;

```
### INSERT INTO `test`.`test_table`
### SET
###   @1=4632662255 /* LONGINT meta=0 nullable=0 is_null=0 */
###   @2=13382498 /* INT meta=0 nullable=0 is_null=0 */
###   @3=536 /* SHORTINT meta=0 nullable=0 is_null=0 */
###   @4=8388607 /* MEDIUMINT meta=0 nullable=0 is_null=0 */
###   @5=3 /* TINYINT meta=0 nullable=0 is_null=0 */
###   @6=1483892218 /* INT meta=0 nullable=0 is_null=0 */
```

```
### INSERT INTO `test`.`test_table`
### SET
###   @1=4632662255 /* LONGINT meta=0 nullable=0 is_null=0 */
###   @2=13382498 /* INT meta=0 nullable=0 is_null=0 */
###   @3=536 /* SHORTINT meta=0 nullable=0 is_null=0 */
###   @4=-6777216 (10000000) /* MEDIUMINT meta=0 nullable=0 is_null=0 */
###   @5=3 /* TINYINT meta=0 nullable=0 is_null=0 */
###   @6=1483892218 /* INT meta=0 nullable=0 is_null=0 */
```

Since the format is different, I guess we hit some kind of bug. This issue looks like similar to this:
https://github.com/github/gh-ost/issues/157 However my version is 1.0.32. MySQL version: Server version: 5.7.10-3-log Percona Server (GPL), Release 3, Revision 63dafaf

The error is;

```
2017-01-08 17:16:03 ERROR Error 1264: Out of range value for column 'column3' at row 1; query=
                        replace /* gh-ost `test`.`test_table` */ into
                                `test`.`test_table`
                                        (`id`, `column1`, `column2`, `column3`, `column4`, `column6`)
                                values
                                        (?, ?, ?, ?, ?, ?)
                ; args=[4632662255, 13382498, 536, 10000000, 3, 1483892218]
```

**Comment by shlomi-noach** [comment]
Thank you,

I don't see that https://github.com/github/gh-ost/issues/348 is related in any way. 
Similar (and old) issue: https://github.com/github/gh-ost/issues/157, fixed by https://github.com/github/gh-ost/pull/160, and in particular by this code: https://github.com/github/gh-ost/pull/160/files#diff-5ba42d73e1ce798d344cb2a18e53497aR35

What I'm missing here, though, is the execution: what's the `alter` statement you're running?

**Comment by shlomi-noach** [comment]
Also, please paste output of `show create table `_test_table_gho`.

Just as FYI, the binary logs always show values as signed; `gh-ost`'s task is to recognize which of those values is truly signed and which is unsigned, and to make to correct conversion.

There's a bunch of integration tests that check for various signed/unsigned scenarios. Once I get requested info I'll try and reproduce via tests.

**Comment by serdarsahin** [comment]
That was the wrong issue, I meant #157. I've updated the original issue description and included the details you requested.

**Comment by shlomi-noach** [comment]
Thank you.
By the way, with regard https://github.com/github/gh-ost/issues/348, please be advised that the fix is merged and released, see https://github.com/github/gh-ost/releases/tag/v1.0.34

**Comment by serdarsahin** [comment]
Ok, I missed that, thought it's already fixed. We'll try this one.

Thanks for the update.

**Comment by shlomi-noach** [comment]
@serdarsahin I think the changes you made to the initial comment caused an inconsistency. Generally speaking, I think adding new comments is preferable to editing comments in-place.

The problem I see now with your edited, first comment: previously the error was on `Out of range value for column 'column4'`, now it is on `Out of range value for column 'column3'`, but I do not see that you've changed the table definition in-place, and the datatype of `column3` (`mediumint`) doesn't make sense to have this error, as opposed to the datatype of `column4` (`tinyint`).

So right now I'm really confused and would appreciate it if you could add a _new_ comment with all the correct, consistent details. Thank you!

**Comment by serdarsahin** [comment]
Hi Shlomi,

Yes, I made a mistake there while anonymising the table and description, that's why I've chosen to update it. It is now correct. We are getting the error for the column3, which is represented in binary logs as "###   @4=-6777216 (10000000) /* MEDIUMINT meta=0 nullable=0 is_null=0 "

If you look at the two binary log output, which I haven't changed, you'll see that only that column is different between two log output.

**Comment by shlomi-noach** [comment]
@serdarsahin your description is unfortunately still inconsistent. The `alter` statement `drop`s a column, and yet the two tables you depict (original/new) have the same number of columns.

I'm writing the tests to reproduce the bug, however the bug itself is inconsistent. Please give this another try.

**Comment by shlomi-noach** [comment]
To further make a suggestion, please don't try and emulate or guess the structure you think is right, but rather provide with actual `show create table test_table`, `show create table _test_table_gho` (if you must anonymize table/column names, please attempt to keep consistency)

**Comment by serdarsahin** [comment]
Hi Shlomi,

The two schema has different number of columns. After I made the editing today, I've added the details and fixed the consistency. For instance, there is no column5 in the second schema. I am sorry, if that created trouble for you, but can't  see the inconsistency right now. Could you be more specific?

**Comment by shlomi-noach** [comment]
@serdarsahin you are right, the tables do make sense -- I must have missed this.

So right now I'm unable to reproduce the problem, and as far as I can see the code handles the case correctly. I realize this doesn't work for you but am a bit lost right now as for why that is the case.

**Comment by shlomi-noach** [comment]
Whoa! Good (bad) news: I'm able to reproduce!

**Comment by shlomi-noach** [comment]
Ah, yes!

The problem is specifically with `mediumint`, being 3-byte, which is unorthodox. More specifically, there is no compatible datatype in `golang` which is why the conversion does not work correctly. Will now explicitly fix this particular use case.

**Comment by shlomi-noach** [comment]
@serdarsahin can you try the attached binary? It should solve the problem with the `mediumint unsigned`:

[gh-ost-binary-linux-20170110095938.tar.gz](https://github.com/github/gh-ost/files/695532/gh-ost-binary-linux-20170110095938.tar.gz)

**Comment by ggunson** [comment]
I've created a tangential issue to this based on the alter statement used in this issue (https://github.com/github/gh-ost/issues/357). 

tl;dr The docs suggest you need a shared unique/PK index between the old and new tables, but in this case the PK becomes a unique key and a new PK is created from a non-unique index, so there is technically no shared index. The new issue is to determine if there are any caveats to that and if we need to update the documentation.

**Comment by shlomi-noach** [comment]
Have edited the title for this Issue.

**Comment by serdarsahin** [comment]
It's been running for more than 22 hours and estimating 12 more hours. So far there are no issues. It looks like it's been fixed but I'll get back to you once it finalises the migration.

**Comment by shlomi-noach** [comment]
How did it go?

**Comment by serdarsahin** [comment]
It's stopped after hitting the critical threshold, which is low (it is taking 70+ hours, so estimation changes depending on the load)

We've started it again and are planning to switch on Monday. Though I can say that it is fixed, as we never come to this point. It was failing after %10 before. 

I'll update this issue again once it finalises.

**Comment by serdarsahin** [comment]
It worked perfectly! Thanks.. 

We've also done data validation row by row and column by column on this table (on 4.5 billion rows).

**Comment by shlomi-noach** [comment]
great, thanks for updating!

**Comment by shlomi-noach** [comment]
@serdarsahin now, regardless of everything, a quick word of advice:

In InnoDB there is no advantage to using `mediumint`. This datatype takes `3` bytes of logical storage. However, internally, InnoDB pads it and uses `4` bytes of storage, same as normal `int`.

FYI

**Comment by ggunson** [comment]
@shlomi-noach Where does it say that InnoDB pads mediumint to 4 bytes? There seems to be disagreement about that: http://www.tocker.ca/2014/05/30/how-important-is-it-to-use-2-byte-and-3-byte-integers.html#comment-1412976626

**Comment by shlomi-noach** [comment]
Hmmm. Pretty sure I read that by Heikki Tuuri some time in the past, but cannot find this anymore. Closest I can find is MariaDB suggesting `MEDIUMINT` is 4 bytes [in memory](https://mariadb.com/kb/en/mariadb/data-type-storage-requirements/).

I'll obviously get @morgo 's word on this, which is authoritative.

**Comment by shlomi-noach** [comment]
fixed by https://github.com/github/gh-ost/pull/355, closing.
A release will be available in a few days.

**Comment by morgo** [comment]
HI!  The test-case in the comments confirms storage is 3 bytes.  Aleks (author of TwinDB recovery tool) also confirmed it on FB after I wrote that post.

I have no reason to dispute InnoDB will use 4 bytes when decoding an individual page in memory.  But this is a pretty limited case; since the full buffer pool in memory is the same as the on-disk representation (except when the punch-hole compression is used or encryption).
